controllers/agentconfig_controller.go

```diff
@@ -0,0 +1,702 @@
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AgentConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&porterv1.AgentConfig{}, builder.WithPredicates(resourceChanged{})).
+		Owns(&porterv1.AgentAction{}).
+		Owns(&corev1.PersistentVolumeClaim{}).
+		Complete(r)
+}
+
+// Reconcile is called when the spec of an agent config is changed
+// or a job associated with an agent config is updated.
+// Either schedule a job to handle a spec change, or update the agent config status in response to the job's state.
+func (r *AgentConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("agent config", req.Name, "namespace", req.Namespace)
+
+	// Retrieve the agent config
+	agentCfg := &porterv1.AgentConfig{}
+	err := r.Get(ctx, req.NamespacedName, agentCfg)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(Log5Trace).Info("Reconciliation skipped: AgentConfig CRD or one of its owned resources was deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log = log.WithValues("resourceVersion", agentCfg.ResourceVersion, "generation", agentCfg.Generation, "observedGeneration", agentCfg.Status.ObservedGeneration)
+	log.V(Log5Trace).Info("Reconciling agent config")
+
+	// Check if we have requested an agent run yet
+	action, handled, err := r.isHandled(ctx, log, agentCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if action != nil {
+		log = log.WithValues("agentaction", action.Name)
+	}
+
+	// Sync the agent config status from the action
+	if err = r.syncStatus(ctx, log, agentCfg, action); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Check if we have finished removing the agentCfg from the pvc and pv's owner reference
+	processed, err := r.isDeleteProcessed(ctx, agentCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if processed {
+		err = removeAgentCfgFinalizer(ctx, log, r.Client, agentCfg)
+		log.V(Log4Debug).Info("Reconciliation complete: Finalizer has been removed from the AgentConfig.")
+		return ctrl.Result{}, err
+	}
+
+	// TODO: once porter has ability to install multiple plugins with one command, we will allow users
+	// to install plugins other than the default one
+	updatedCfg := setDefaultPlugins(agentCfg)
+	if updatedCfg {
+		err := r.Update(ctx, agentCfg)
+		return ctrl.Result{}, err
+	}
+
+	readyPVC, tempPVC, err := r.GetPersistentVolumeClaims(ctx, log, agentCfg)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if readyPVC != nil && tempPVC == nil && !isDeleted(agentCfg) {
```

Comment:
I recommend moving all of this into multiple functions. If you look over the other reconcile functions in the other controller, I've tried real hard to keep them short and readable because otherwise this function is overwhelming and difficult to test/maintain.